### PR TITLE
*: Initialize gRPC server metrics

### PR DESCRIPTION
**Pull request by kakkoyun**
This PR initializes gRPC server metrics as soon as server starts, to provide more reliable data for instrumentation.

## Changes

* Initialize gRPC server metrics

## Verification

* `make test | build | lint` and local test

### CODE REVIEW

File: cmd/thanos/query.go [hunk 1]
```diff
@@ -441,13 +441,15 @@ func runQuery(
 		}
 		logger := log.With(logger, "component", component.Query.String())
 
-		opts, err := defaultGRPCServerOpts(logger, reg, tracer, srvCert, srvKey, srvClientCA)
+		met := grpc_prometheus.NewServerMetrics()
+		opts, err := defaultGRPCServerOpts(logger, reg, tracer, met, srvCert, srvKey, srvClientCA)
```
Comment (brancz) [inline cmd/thanos/query.go]: We should append the options afterwards I think, instead of passing them in here.

File: cmd/thanos/main.go [hunk 2]
```diff
@@ -248,8 +248,7 @@ func registerMetrics(mux *http.ServeMux, g prometheus.Gatherer) {
 // - request histogram
 // - tracing
 // - panic recovery with panic counter
-func defaultGRPCServerOpts(logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, cert, key, clientCA string) ([]grpc.ServerOption, error) {
-	met := grpc_prometheus.NewServerMetrics()
+func defaultGRPCServerOpts(logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, met *grpc_prometheus.ServerMetrics, cert, key, clientCA string) ([]grpc.ServerOption, error) {
```
Comment (brancz) [inline cmd/thanos/main.go]: My understanding is we wanted to append the metrics as options but missed to do so here. I’d suggest we do it outside of this function anyways.

File: cmd/thanos/query.go (hunk 1 above)
Comment (kakkoyun) [reply cmd/thanos/query.go]: Could please clarify more. Should I extract instrumentation options out from this method and append them later?

**Comment by kakkoyun** [comment]
It's failed because of a dependency issue:
```console
>> fetching github.com/minio/minio@RELEASE.2018-10-06T00-15-16Z revision/version
# cd .; git clone https://git.apache.org/thrift.git /tmp/thanos-go/src/git.apache.org/thrift.git
Cloning into '/tmp/thanos-go/src/git.apache.org/thrift.git'...
fatal: unable to access 'https://git.apache.org/thrift.git/': Failed to connect to git.apache.org port 443: Connection timed out
package github.com/minio/minio
	imports git.apache.org/thrift.git/lib/go/thrift: exit status 128
Makefile:209: recipe for target 'test-deps' failed
make[2]: *** [test-deps] Error 1
make[2]: Leaving directory '/go/src/github.com/thanos-io/thanos'
Makefile:193: recipe for target 'test-only-gcs' failed
make[1]: *** [test-only-gcs] Error 2
make[1]: Leaving directory '/go/src/github.com/thanos-io/thanos'
Makefile:202: recipe for target 'test-local' failed
make: *** [test-local] Error 2
Exited with code 2
```

**Comment by kakkoyun** [comment]
There's already an issue (#1483) opened for the failure.

**Pull request merged**